code/processes/gateway.q

```diff
@@ -284,8 +284,13 @@ checktimeout:{
   finishquery[qids;1b;0Ni]];
  }
 
-
-
+// check if hdb query has date
+checkdate:{[query;servertype]
+ if[servertype=`hdb;
+  $[type [query] = 0h;
+  if[not any raze[string query 2] like "*date*"; '"query requires a date"]; 
```

Comment:
This isn't very nice, and it shouldn't be necessary to convert to a string...In fact, I think you'd probably be best to *not* work with strings at all....

```
q)0N!parse"select from trades where date=2018.01.01";
(?;`trades;,,(=;`date;2018.01.01);0b;())
q)q:(?;`trades;enlist(=;`date;2018.01.01);0b;())
q)q[2]
= `date 2018.01.01
q)any `date in/:q[2]
1b
```

Something like this will work if the query is passed as a string (in which case, `parse` it) or a functional form list (in which case, don't parse it)

Of course, you also need to know what type of partition the HDB is using - this is stored in `.Q.pf` http://code.kx.com/q/ref/dotq/#qpf-partition-type

Finally, you also need to handle cases where the client is calling a predefined function, rather than a Q-SQL query. E.g.

```
.gw.syncexec["vwap[`AAPL;2018.01.01]";`hdb]
```

Again, parsing this string may help...

```
q)parse"vwap[`AAPL;2018.01.01]"
`vwap
,`AAPL
2018.01.01
```

(of course, the string here doesn't contain the word `"date"`, but should absolutely be allowed to run)